File: file-templates/templates/MVVM/List.xctemplate/UIViewControllerObjective-C/___FILEBASENAME___ViewController.m
//___FILEHEADER___

#import "___FILEBASENAME___.h"
#import "___VARIABLE_productName___ViewModel.h"
#import "___VARIABLE_productName___View.h"
#import <Masonry/Masonry.h>

@interface ___VARIABLE_productName___ViewController ()<___VARIABLE_productName___ViewModelInteractionProvider>
@property (nonatomic, strong) ___VARIABLE_productName___ViewModel *viewModel;
@property (nonatomic, strong) ___VARIABLE_productName___View *mainView;
@end

@implementation ___VARIABLE_productName___ViewController

- (instancetype)initWithViewModel:(___VARIABLE_productName___ViewModel *)viewModel {
    self = [super init];
    if (self) {
        self.viewModel = viewModel;
        self.viewModel.interactionProvider = self;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];

}

#pragma mark - ___VARIABLE_productName___ViewModelInteractionProvider

- (UIViewController *)interactionViewController {
    return self;
}

#pragma mark - getter
- (___VARIABLE_productName___View *)mainView {
    if (!_mainView) {
        ___VARIABLE_productName___View *view = [[___VARIABLE_productName___View alloc] initWithViewModel:self.viewModel];
        [self.view addSubview:view];
        [view mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(self.view.mas_safeAreaLayoutGuideTop).offset(0);
            make.left.equalTo(self.view.mas_left);
            make.right.equalTo(self.view.mas_right);
            make.bottom.equalTo(self.view.mas_safeAreaLayoutGuideBottom).offset(0);
        }];
        _mainView = view;
    }
    return _mainView;
}

@end

File: file-templates/templates/MVVM/MVVM.xctemplate/UIViewControllerSwift/___FILEBASENAME___ViewController.swift
//  ___FILEHEADER___
//  Summary: ___VARIABLE_Summary___

import UIKit
import SnapKit

final class ___VARIABLE_productName:identifier___ViewController: UIViewController {
    
    // MARK: - Private properties
    
    private let viewModel: ___VARIABLE_productName:identifier___ViewModel
    
    lazy var mainView: ___VARIABLE_productName:identifier___View = {
        let mainView = ___VARIABLE_productName:identifier___View(viewModel: self.viewModel)
        return mainView
    }()

    // MARK: - Init

    init(viewModel: ___VARIABLE_productName:identifier___ViewModel) {
        self.viewModel = viewModel
        super.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    // MARK: override
    
    func addUIComponents() {
      self.view.addSubview(mainView)
    }
    
    func layoutUIComponents() {
        mainView.snp.makeConstraints { make in
            make.top.equalTo(self.view.safeAreaLayoutGuide.snp.top)
            make.left.equalTo(self.view.snp.left)
            make.right.equalTo(self.view.snp.right)
            make.bottom.equalTo(self.view.safeAreaLayoutGuide.snp.bottom)
        }
    }
    
    func setupBindings() {
         
    }
}